File: alice/gamma/sdk/python/gamma_sdk/sdk/button.py
# coding:utf-8
import attr
import json

import alice.gamma.sdk.api.api_pb2 as api


@attr.s(frozen=True)
class Button:
    title = attr.ib(type=str)
    url = attr.ib(type=str, default=None)
    payload = attr.ib(default=None)
    hide = attr.ib(type=bool, default=False)

    def to_proto(self):
        return api.Button(
            title=self.title,
            url=self.url,
            payload=json.dumps(self.payload, ensure_ascii=False).encode(encoding='utf-8') if self.payload else None,
            hide=self.hide,
        )

    @staticmethod
    def buttons_to_proto(buttons):
        return [button.to_proto() for button in buttons]

File: alice/gamma/sdk/golang/button.go
package sdk

import (
	"encoding/json"

	"a.yandex-team.ru/alice/gamma/sdk/api"
)

type Button struct {
	Title   string
	Payload interface{}
	URL     string
	Hide    bool
}

func (button *Button) toProto() (protoButton *api.Button, err error) {
	protoButton = &api.Button{
		Title: button.Title,
		Url:   button.URL,
		Hide:  button.Hide,
	}
	if button.Payload != nil {
		protoButton.Payload, err = json.Marshal(button.Payload)
		if err != nil {
			return nil, err
		}
	}
	return protoButton, nil
}

func buttonsToProto(buttons []Button) (protoButtons []*api.Button, err error) {
	protoButtons = make([]*api.Button, len(buttons))
	for i, button := range buttons {
		protoButtons[i], err = button.toProto()
		if err != nil {
			return protoButtons, err
		}
	}
	return protoButtons, nil
}
